dali/ft/filecopy.cpp

```diff
@@ -3249,13 +3256,117 @@ void FileSprayer::updateTargetProperties()
+                    curHistory->addPropTree(aname, createPTreeFromIPT(&historyIter->query()));
+                }
+            }
+
+            // Add new record about this operation
+            IPropertyTree * newRecord = createPTree("Origin");
+
+            CDateTime temp;
+            temp.setNow();
+            unsigned hour, min, sec, nanosec;
+            temp.getTime(hour, min, sec, nanosec);
+            temp.setTime(hour, min, sec, 0);
+
+            StringBuffer timestr;
+            newRecord->addProp("@timestamp",temp.getString(timestr).str());
+
+            StringBuffer val;
+            srcAttr->getProp("@owner", val);
+            newRecord->addProp("@owner", val.str());
+
+            srcAttr->getProp("@workunit", val.clear());
+            if (0 == val.length())
+                val.set("N/A");
+            newRecord->addProp("@workunit", val.str());
+
+            newRecord->addProp("@operation", getOperationTypeString());
+
+            // add original file name
+            if (distributedSource)
+            {
+                RemoteFilename remoteFile;
+                distributedSource->getPart(0)->getFilename(remoteFile, 0 );
+                splitAndStoreFileInfo(newRecord, remoteFile);
+            }
+            else
+            {
+                val.clear();
+                ForEachItemIn(idx, sources)
+                {
+                    FilePartInfo & curSource = sources.item(idx);
+                    RemoteFilename remoteFile = curSource.filename;
+                    splitAndStoreFileInfo(newRecord, remoteFile, idx, false);
+                }
+            }
+            curHistory->addPropTree("Origin",newRecord);
         }
     }
     if (error)
         throw error.getClear();
 }
 
+void FileSprayer::splitAndStoreFileInfo(IPropertyTree * newRecord, RemoteFilename &remoteFileName
+                                      , aindex_t idx, bool isDistributedSource)
+{
+    StringBuffer drive;
+    StringBuffer path;
+    StringBuffer fileName;
+    StringBuffer ext;
+    remoteFileName.split(&drive, &path, &fileName, &ext);
+    if (idx == 0)
+    {
+        if(drive.isEmpty())
+        {
+            remoteFileName.queryIP().getIpText(drive.clear());
+            newRecord->addProp("@ip", drive.str());
+        }
+        else
+            newRecord->addProp("@drive", drive.str());
+
+        newRecord->addProp("@path", path.str());
+    }
+    if (!isDistributedSource && ext.length())
+        fileName.append(ext);
+
+    // In spray multiple source files case keep all original filenames
+    if (newRecord->hasProp("@name"))
+    {
+        StringBuffer currentName;
+        newRecord->getProp("@name", currentName);
+        currentName.append(",").append(fileName);
+        fileName = currentName;
+    }
 
+    newRecord->addProp("@name", fileName.str());
+}
+
+void FileSprayer::setOperation(dfu_operation op)
+{
+    operation = op;
+}
+
+dfu_operation FileSprayer::getOperation(void)
```

Comment:
style: we don't tend to prototype methods as (void).

Also could be const method.